fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/AvroDimensionRowParser.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Parses an AVRO file into Dimension Rows.
+ */
+public class AvroDimensionRowParser {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AvroDimensionRowParser.class);
+    private ObjectMapper objectMapper;
+
+    private final DimensionFieldNameMapper dimensionFieldNameMapper;
+
+    /**
+     * Constructs an AvroDimensionRowParser object based on the DimensionFieldNameMapper object.
+     *
+     * @param dimensionFieldNameMapper Object that defines the dimension field name transformations
+     * @param objectMapper Object that is used to construct mapper objects
+     */
+    public AvroDimensionRowParser(DimensionFieldNameMapper dimensionFieldNameMapper, ObjectMapper objectMapper) {
+        this.dimensionFieldNameMapper = memoize(dimensionFieldNameMapper);
+        this.objectMapper = objectMapper;
+    }
+
+    /**
+     * Validates the schema of the given AVRO file with the Dimension schema configured by the user.
+     *
+     * <p>
+     * The sample schema is expected to be in the following format
+     * <pre><code>
+     * {
+     *    "type" : "record",
+     *    "name" : "TUPLE_0",
+     *    "fields" : [
+     *    {
+     *    "name" : "FOO_ID",
+     *    "type" : [ "null","int" ]
+     *    },
+     *    {
+     *    "name" : "FOO_DESC",
+     *    "type" : [ "null", "string" ]
+     *    }
+     *    ]
+     * }
+     * </code></pre>
+     *
+     * @param dimension The dimension object used to configure the dimension
+     * @param avroSchemaPath The path of the AVRO schema file (.avsc)
+     *
+     * @return true if the schema is valid, false otherwise
+     *
+     * @throws IllegalArgumentException thrown if JSON object `fields` is not present
+     */
+    private boolean doesSchemaContainAllDimensionFields(Dimension dimension, String avroSchemaPath)
+        throws IllegalArgumentException {
+
+        // Convert the AVRO schema file into JsonNode Object
+        JsonNode jsonAvroSchema = convertFileToJsonNode(avroSchemaPath);
+
+        try {
+            // The following method throws a NullPointerException, if the JSON schema doesn't have the `fields` object
+            // Avoiding the `Optional.of().orElseThrow()` to log the error details.
+            Optional.of(jsonAvroSchema.get("fields"));
+        } catch (NullPointerException e) {
+            String msg = "`fields` is a required JSON field in the avro schema {} ";
+            LOG.error(msg, e);
+            throw new IllegalArgumentException(msg);
```

Comment:
If we catch and throw a new exception, we should always be sure to include the initiating exception as the cause so that we don't lose the full trace.